src/traits/data.jl

```diff
@@ -136,17 +136,124 @@ Tables.materializer(D::Type{<:Data}) = D
 # DATAFRAME INTERFACE
 # --------------------
 
-function Base.getproperty(data::Data, col::Symbol)
-  if col == :geometry
+function Base.getproperty(data::Data, var::Symbol)
+  if var == :geometry
     domain(data)
   else
     cols = Tables.columns(values(data))
-    Tables.getcolumn(cols, col)
+    Tables.getcolumn(cols, var)
   end
 end
 
-Base.getproperty(data::Data, col::AbstractString) =
-  getproperty(data, Symbol(col))
+Base.getproperty(data::Data, var::AbstractString) =
+  getproperty(data, Symbol(var))
+
+function Base.getindex(data::Data, ind::Int, vars::AbstractVector{Symbol})
+  _checkvars(vars)
+  _rmgeometry!(vars)
+  dom   = domain(data)
+  table = values(data)
+  row   = Tables.subset(table, ind)
+  pairs = (var => Tables.getcolumn(row, var) for var in vars)
+  (; pairs..., geometry=dom[ind])
+end
+
+Base.getindex(data::Data, ind::Int, vars::AbstractVector{<:AbstractString}) =
+  getindex(data, ind, Symbol.(vars))
+
+function Base.getindex(data::Data, ind::Int, ::Colon)
+  dom   = domain(data)
+  table = values(data)
+  row   = Tables.subset(table, ind)
+  vars  = Tables.columnnames(row)
+  pairs = (var => Tables.getcolumn(row, var) for var in vars)
+  (; pairs..., geometry=dom[ind])
+end
+
+Base.getindex(data::Data, ind::Int, var::Symbol) =
+  getproperty(data, var)[ind]
+
+Base.getindex(data::Data, ind::Int, var::AbstractString) =
+  getindex(data, ind, Symbol(var))
+
+function Base.getindex(data::Data, inds::AbstractVector{Int}, vars::AbstractVector{Symbol})
+  _checkvars(vars)
+  _rmgeometry!(vars)
+  table = values(data)
+
+  newdom = view(domain(data), inds)
+  subset = Tables.subset(table, inds)
+  cols   = Tables.columns(subset)
+  𝒯 = NamedTuple(var => Tables.getcolumn(cols, var) for var in vars)
+  newtable = 𝒯 |> Tables.materializer(table)
+
+  vals = Dict(paramdim(newdom) => newtable)
+  constructor(data)(newdom, vals)
+end
+
+Base.getindex(data::Data, inds::AbstractVector{Int}, vars::AbstractVector{<:AbstractString}) =
+  getindex(data, inds, Symbol.(vars))
+
+function Base.getindex(data::Data, inds::AbstractVector{Int}, ::Colon)
+  table = values(data)
+
+  newdom   = view(domain(data), inds)
+  newtable = Tables.subset(table, inds)
+
+  vals = Dict(paramdim(newdom) => newtable)
+  constructor(data)(newdom, vals)
+end
+
+function Base.getindex(data::Data, inds::AbstractVector{Int}, var::Symbol)
+  if var == :geometry
+    view(domain(data), inds)
+  else
+    table  = values(data)
+    subset = Tables.subset(table, inds)
+    cols   = Tables.columns(subset)
+    Tables.getcolumn(cols, var)
+  end
+end
+
+Base.getindex(data::Data, inds::AbstractVector{Int}, var::AbstractString) =
+  getindex(data, inds, Symbol(var))
+
+function Base.getindex(data::Data, ::Colon, vars::AbstractVector{Symbol})
+  _checkvars(vars)
+  _rmgeometry!(vars)
+  dom   = domain(data)
+  table = values(data)
+
+  cols = Tables.columns(table)
+  𝒯 = NamedTuple(var => Tables.getcolumn(cols, var) for var in vars)
```

Comment:
let's rename T by pairs and use (; pairs...) instead of explicitly writing NamedTuple